### PR TITLE
Add upper_triangle and docs for {Log,Root}DetCone{Triangle,Square}

### PR DESCRIPTION
**Pull request by odow**
Replacement for #3427

### CODE REVIEW

**Comment by blegat** [comment]
It could also be the lower triangle row wise. Maybe simply `triangle_vec` to be similar to `vec`.

File: docs/src/tutorials/conic/tips_and_tricks.jl [hunk 1]
```diff
@@ -338,6 +338,78 @@ set_silent(model)
...
+# When forming the function, use [`triangle_vec`](@ref) to obtain the upper
+# triangle of the matrix as a vector in the order that JuMP requires.
+
+model = Model(SCS.Optimizer)
+set_silent(model)
+@variable(model, t)
+@variable(model, u)
+@variable(model, X[1:2, 1:2], Symmetric)
+@objective(model, Max, t)
+@constraint(model, [t; u; triangle_vec(X)] in MOI.LogDetConeTriangle(2))
```
Comment (odow) [inline docs/src/tutorials/conic/tips_and_tricks.jl:407]: I've updated to `triangle_vec`. I think this function + this part of the documentation is enough to fix the previously ugly syntax for using `{Log,Root}DetConeTriangle`.

File: docs/src/tutorials/conic/tips_and_tricks.jl [hunk 2]
```diff
@@ -338,6 +338,78 @@ set_silent(model)
...
+@variable(model, t)
+@variable(model, X[1:2, 1:2])
+@objective(model, Max, t)
+@constraint(model, [t; vec(X)] in MOI.RootDetConeSquare(2))
+@constraint(model, X .== [2 1; 1 3])
+optimize!(model)
+value(t), sqrt(LinearAlgebra.det(value.(X)))
+
+# If `X` is symmetric, then you can use [`MOI.RootDetConeTriangle`](@ref)
+# instead. This can be more efficient because the solver does not need to add
```
Comment (mlubin) [inline docs/src/tutorials/conic/tips_and_tricks.jl:359]: Can be or is more efficient?

**Review by mlubin — APPROVED** [review]

File: docs/src/tutorials/conic/tips_and_tricks.jl (hunk 2 above)
Comment (odow) [reply docs/src/tutorials/conic/tips_and_tricks.jl:359]: Can be.

It's a bit subtle: the Square->Triangle bridge will only add the equality constraints for elements that are not equivalent using `==` (or really, `isapprox`).

```julia
model = Model()
@variable(model, t)
@variable(model, x)
@variable(model, y)
@variable(model, z)

# Elements are symmetric, no constraints added 
@constraint(model, [t; vec([x y; y z])] in MOI.RootDetConeSquare(2))

# Elements are symmetric, constraints added, even though there exists a
# separate `y == w` constraint
@variable(model, w)
@constraint(model, y == w)
@constraint(model, [t; vec([x w; y z])] in MOI.RootDetConeSquare(2))
```

File: docs/src/tutorials/conic/tips_and_tricks.jl [hunk 3]
```diff
@@ -338,6 +338,78 @@ set_silent(model)
...
+# additional constraints to ensure `X` is symmetric.
+
+# When forming the function, use [`triangle_vec`](@ref) to obtain the upper
```
Comment (blegat) [inline docs/src/tutorials/conic/tips_and_tricks.jl:362]: I'd clarify by saying "column-wise upper triangle"

**Pull request merged**